### DISCUSSION TITLE
A question from api-routes example

### DISCUSSION THREAD

**Discussion by arunppsg** [vercel/next.js#44516] [Help] [answered]
I am learning about nextjs/typescript and I noticed that when this line is removed,

https://github.com/vercel/next.js/blob/00d7a35ed79b34042b0d102f47fc5df35db7f90a/examples/api-routes/pages/person/%5Bid%5D.tsx#L22

the app throws an error `Uncaught TypeError: data is undefined`. I am not able to understand why it throws the error and why that line is required.

When I removed the second table row (the {data.name}, {data.height}... lines) and do a `console.log(data)`, it shows the data object Object { id: "1", name: "Luke Skywalker", height: "172", mass: "77", hair_color: "blond", skin_color: "fair", eye_color: "blue", gender: "male" }
 object. This makes it more confusing - data is getting fetched but not rendered.

**Comment by hubertlepicki** [comment, marked as answer]
So what happens is that on this line, you're fetching the data:

https://github.com/vercel/next.js/blob/00d7a35ed79b34042b0d102f47fc5df35db7f90a/examples/api-routes/pages/person/%5Bid%5D.tsx#L16

This actually doesn't happen instantly, as the request is made to fetch the data. While the request is being made, and data is being loaded, your component doesn't wait for it. It goes on with it's first render cycle, under which `data` is indeed not loaded. It renders the page that says "Loading..." very briefly, maybe even so quick you cannot notice it. This is when there's no data loaded yet. Once data loads, React will re-render the component with data as loaded from the API.

**Comment by icyJoseph** [comment]
Asynchronous operations such as fetching data from an endpoint, have two characteristics: 

- unknown duration
- they may fail

Since React components, leaving aside Server Components, are synchronous, interfacing with an asynchronous operation, in this case a promise, requires a few considerations. Not exhaustive list:

- What to show while we wait for data
- What to show when we fail to get the data
- What to do if we have data

So as the React component is rendered, SWR is hit, and if there's no data for the given key, while SWR fetches it, you are given back no data for the render frame. Later on when the data arrives, a new render frame happens, and now data has the value SWR got from the network, or an error!

**Comment by arunppsg** [comment]
A related question: since data is not immediately returned, I should a `if (!data)` clause after every useSWR right?

**Reply by icyJoseph** [reply]
Well, it depends, https://github.com/vercel/swr/discussions/1408#discussioncomment-1383367:

You could write a custom hook that calculates whether it is loading or not, and use that instead. 

The thing is that your component might depend on data in various degrees:

```tsx
const {data} = useSWR(/**/)

return (
  <div>
   <h1>Hello</h1>
   <p>You name is {data?.name ?? ""}</p>
  </div>
)
```

This is why I was a bit more abstract with my answer, and said, `What to show while we wait for data`, instead of, straight up showing a different piece of UI.

Do remember that the `if (loading) return SomethingElse` pattern might cause Layout shift, and prevent other parts of the UI that do not depend on the data directly from showing.

However it is possible that you want to calculate, or derive more data from the `useSWR` result, and in such case, you'd need to guard that the data is no undefined, either with an early exit, a custom hook, or a wrapper component.

**Reply by arunppsg** [reply]
Thanks, this helps me to understand better.
